lib/charms/opensearch/v0/opensearch_plugin_manager.py

```diff
@@ -222,30 +279,43 @@ def _user_requested_to_enable(self, plugin: OpenSearchPlugin) -> bool:
         return True
 
     def _is_enabled(self, plugin: OpenSearchPlugin) -> bool:
-        """Returns true if plugin is enabled."""
-        # If not requested to be disabled, check if options are configured or not
-        are_configs_enabled = plugin.config().config_entries_to_add
-        stored_plugin_conf = self._opensearch_config.get_plugin(are_configs_enabled)
+        """Returns true if plugin is enabled.
 
-        if not stored_plugin_conf:
-            return False
+        The main question to answer is if we would remove the configuration from
+        opensearch.yaml and/or add some other configuration instead.
+        If yes, then we know that the service is enabled.
 
-        for key, val in stored_plugin_conf.items():
-            if are_configs_enabled.get(key) != val:
-                return False
-        return True
+        The disable() method is used, as a given entry will be removed from opensearch.yml.
+
+        If disable() returns a list, then check if the keys in the stored_plugin_conf
+        matches the list values; otherwise, if a dict, then check if the keys and values match.
+        If they match in any of the cases above, then return True.
+        """
+        try:
+            plugin_conf = plugin.disable().config_entries_to_del
+            stored_plugin_conf = self._opensearch_config.get_plugin(plugin_conf)
+            # Using sets to guarantee matches; stored_plugin_conf will be a dict
+            if isinstance(plugin_conf, list):
+                return set(plugin_conf) == set(stored_plugin_conf)
+            else:  # plugin_conf is a dict
+                return plugin_conf == stored_plugin_conf
+        except (KeyError, OpenSearchPluginError) as e:
+            logger.warning(f"_is_enabled: error with {e}")
+            return False
 
     def _needs_upgrade(self, plugin: OpenSearchPlugin) -> bool:
         """Returns true if plugin needs upgrade."""
-        current_version = plugin.version
-        return self._opensearch.version != ".".join(current_version.split(".")[0:3])
+        plugin_version = plugin.version.split(".")
+        version = self._opensearch.version.split(".")
+        num_points = min(len(plugin_version), len(version))
+        return version[:num_points] != plugin_version[:num_points]
 
     def _is_plugin_relation_set(self, relation_name: str) -> bool:
         """Returns True if a relation is expected and it is set."""
         if not relation_name:
             return False
         relation = self._charm.model.get_relation(relation_name)
-        return relation is not None
+        return relation is not None and relation.units
```

Comment:
Note! `relation-created`, `relation-broken`, `relation-disabled` and I'm a bit unsure about `relation-joined` (if the counter is already 1 from zero or not yet...?)

Reply:
Hi @juditnovak added more details in the docstring of `opensearch_plugins`, in this PR: https://github.com/canonical/opensearch-operator/pull/137/files#diff-b0054790c1d68243f1b71a3094cf7f0f989456fd7de12f17a92fd1b4c7ca30eb

Reply:
Hmmm.... but if we recommend doesn't change the fact that adding the `units` number here will make `_is_plugin_relation_set() == False` within the `PluginManager.run()` execution.

This has consequences on the workflow if invoked in a `RelationCreated` event...

Reply:
@juditnovak implemented the `event-scope`, as we've discussed.